File: embed.py
import cv2
import numpy as np
import random
from pyzbar.pyzbar import decode, ZBarSymbol  # Add this import

def get_qr_ec_level(qr_path):
    # """Detect QR code error correction level"""
    try:
        decoded = decode(cv2.imread(qr_path), symbols=[ZBarSymbol.QRCODE])
        if decoded:
            # EC levels: L=0, M=1, Q=2, H=3
            return decoded[0].quality
        return 1  # Default to M if detection fails
    except:
        return 1  # Default to M if library fails

def calculate_max_embed_size(qr_size, ec_level):
    # """Calculate safe embed size based on error correction"""
    total_modules = qr_size ** 2
    if ec_level == 0:    # L-level (7%)
        return int(np.sqrt(0.07 * total_modules))
    elif ec_level == 1:  # M-level (15%)
        return int(np.sqrt(0.15 * total_modules))
    else:                # Q/H levels (25-30%)
        return qr_size // 3  # Original behavior

def is_safe_zone(qr_img, x, y, w, h, module_size):
    finder_size = 7 * module_size
    quiet_zone = 4 * module_size

    if ((x < finder_size + quiet_zone and y < finder_size + quiet_zone) or
        (x > qr_img.shape[1] - finder_size - quiet_zone and y < finder_size + quiet_zone) or
        (x < finder_size + quiet_zone and y > qr_img.shape[0] - finder_size - quiet_zone)):
        return False

    timing_pos = finder_size + quiet_zone - module_size // 2
    if abs(y - timing_pos) < module_size // 2 + 1 or abs(x - timing_pos) < module_size // 2 + 1:
        return False

    return (x + w <= qr_img.shape[1] and y + h <= qr_img.shape[0])

def blend_edges_with_qr(qr_img, embed_img, x, y, module_size, blend_percent):
    border_width = module_size * 2
    # scale_alpha = max(0, min(100, blend_percent)) / 100.0  # C++-style clamping
    scale_alpha = np.clip(blend_percent / 100.0, 0.0, 1.0)
    
    # C++-style color sampling
    top_color = np.mean(qr_img[max(0,y-1):y, x:x+embed_img.shape[1]], axis=(0,1))
    bottom_color = np.mean(qr_img[min(qr_img.shape[0]-1,y+embed_img.shape[0]):y+embed_img.shape[0]+1, x:x+embed_img.shape[1]], axis=(0,1))
    left_color = np.mean(qr_img[y:y+embed_img.shape[0], max(0,x-1):x], axis=(0,1))
    right_color = np.mean(qr_img[y:y+embed_img.shape[0], min(qr_img.shape[1]-1,x+embed_img.shape[1]):x+embed_img.shape[1]+1], axis=(0,1))
    
    target_color = (top_color + bottom_color + left_color + right_color) / 4  # C++ averaging

    blend_mask = np.zeros(embed_img.shape[:2], dtype=np.float32)
    h, w = embed_img.shape[:2]

    for i in range(h):
        for j in range(w):
            top_alpha = 1.0 - min(i, border_width) / border_width
            bottom_alpha = 1.0 - min(h-1-i, border_width) / border_width
            left_alpha = 1.0 - min(j, border_width) / border_width
            right_alpha = 1.0 - min(w-1-j, border_width) / border_width
            blend_mask[i,j] = max(top_alpha, bottom_alpha, left_alpha, right_alpha) * scale_alpha

    roi = qr_img[y:y+h, x:x+w].astype(np.float32)
    embed = embed_img.astype(np.float32)

    if embed.shape[2] == 4:
        alpha = embed[:, :, 3] / 255.0
        for c in range(3):
            # C++-style blending with target_color
            # roi[:, :, c] = blend_mask * alpha * embed[:, :, c] + (1 - blend_mask * alpha) * target_color[c]
            roi[:, :, c] = ((1 - blend_mask) * alpha * embed[:, :, c] + blend_mask * target_color[c])
    else:
        for c in range(3):
            # roi[:, :, c] = blend_mask * embed[:, :, c] + (1 - blend_mask) * target_color[c]
            roi[:, :, c] = ((1 - blend_mask) * embed[:, :, c] + blend_mask * target_color[c])

    qr_img[y:y+h, x:x+w] = np.clip(roi, 0, 255).astype(np.uint8)

def embed_image(qr_path, embed_path, output_path, blend_percent, seed=''):
    qr_img = cv2.imread(qr_path, cv2.IMREAD_COLOR)
    embed_img = cv2.imread(embed_path, cv2.IMREAD_UNCHANGED)
    ec_level = get_qr_ec_level(qr_path) #added 
    
    if qr_img is None or embed_img is None:
        print("Failed to read input images.")
        return

    module_size = max(1, qr_img.shape[1] / 21)  # Match C++ integer division

    # Calculate safe embed size based on EC level
    qr_size = min(qr_img.shape[:2]
    max_dim = calculate_max_embed_size(qr_size, ec_level)
                  
    # max_dim = min(qr_img.shape[:2]) // 3
    h, w = embed_img.shape[:2]
    scale = min(max_dim / h, max_dim / w, 1.0)
    embed_img = cv2.resize(embed_img, (int(w * scale), int(h * scale)), interpolation=cv2.INTER_AREA)

    w, h = embed_img.shape[1], embed_img.shape[0]

    # C++-style RNG initialization
    combined_seed = str(random.random()) + seed
    seed_hash = 0
    for c in combined_seed:
        seed_hash = (seed_hash << 5) + seed_hash + ord(c)  # Identical to C++

    rng_x = random.Random(seed_hash)
    rng_y = random.Random(seed_hash ^ 0x55555555)  # Same XOR pattern

    for _ in range(100):
        x = rng_x.randint(0, qr_img.shape[1] - w)
        y = rng_y.randint(0, qr_img.shape[0] - h)
        if is_safe_zone(qr_img, x, y, w, h, module_size):
            break
    else:
        x = min(qr_img.shape[1] - w, max(0, x))
        y = min(qr_img.shape[0] - h, max(0, y))

    blend_edges_with_qr(qr_img, embed_img, x, y, module_size, blend_percent)
    cv2.imwrite(output_path, qr_img)
    print(f"Saved embedded image at {output_path}")

if __name__ == "__main__":
    import sys
    if len(sys.argv) < 6:
        print("Usage: python embed.py <qrPath> <embedPath> <outputPath> <seed> <blendPercent>")
    else:
        blend_percent = max(0, min(100, int(sys.argv[5])))  # C++-style clamping
        embed_image(sys.argv[1], sys.argv[2], sys.argv[3], blend_percent, sys.argv[4])

File: embed.cpp
#include <opencv4/opencv2/core.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>
#include <iostream>
#include <random>

using namespace cv;
using namespace std;

// Function to check safe zones (simplified without ZBar)
bool isSafeZone(const Mat& qrImg, int x, int y, int size, int moduleSize) {
    // QR code structure constants
    int finderPatternSize = 7 * moduleSize;
    int quietZone = 4 * moduleSize;
    
    // Check finder patterns
    if ((x < finderPatternSize + quietZone && y < finderPatternSize + quietZone) ||
        (x > qrImg.cols - finderPatternSize - quietZone && y < finderPatternSize + quietZone) ||
        (x < finderPatternSize + quietZone && y > qrImg.rows - finderPatternSize - quietZone)) {
        return false;
    }
    
    // Check timing patterns
    int timingPos = finderPatternSize + quietZone - moduleSize/2;
    if (abs(y - timingPos) < moduleSize/2 + 1 || abs(x - timingPos) < moduleSize/2 + 1) {
        return false;
    }
    
    // Check bounds
    return !(x + size > qrImg.cols || y + size > qrImg.rows);
}

void blendEdgesWithQR(Mat& qrImg, Mat& embedImg, int x, int y, int moduleSize, int blendPercent ) //int blendPercent added
{
    int borderWidth = moduleSize * 2;
    double scaleAlpha = max(0, min(100, blendPercent)) / 100.0; //edited: this line added
    
    // Sample nearby QR colors
    Scalar topColor = mean(qrImg(Rect(x, max(0,y-1), embedImg.cols, 1)));
    Scalar rightColor = mean(qrImg(Rect(min(qrImg.cols-1,x+embedImg.cols), y, 1, embedImg.rows)));
    Scalar bottomColor = mean(qrImg(Rect(x, min(qrImg.rows-1,y+embedImg.rows), embedImg.cols, 1)));
    Scalar leftColor = mean(qrImg(Rect(max(0,x-1), y, 1, embedImg.rows)));
    
    // Create blend mask
    Mat blendMask = Mat::zeros(embedImg.rows, embedImg.cols, CV_32FC1);
    
    // Horizontal gradients
    for (int i = 0; i < embedImg.rows; i++) {
        float topAlpha = 1.0f - (float)min(i, borderWidth)/borderWidth;
        float bottomAlpha = 1.0f - (float)min(embedImg.rows-1-i, borderWidth)/borderWidth;
        float rowAlpha = max(topAlpha, bottomAlpha);
        
        for (int j = 0; j < embedImg.cols; j++) {
            float leftAlpha = 1.0f - (float)min(j, borderWidth)/borderWidth;
            float rightAlpha = 1.0f - (float)min(embedImg.cols-1-j, borderWidth)/borderWidth;
            blendMask.at<float>(i,j) = max(rowAlpha, max(leftAlpha, rightAlpha));
        }
    }

    // Apply blend percentage scaling
    blendMask *= scaleAlpha;  //eited: this segment added

    
    // Get ROI
    Rect roiRect(x, y, embedImg.cols, embedImg.rows);
    Mat qrROI = qrImg(roiRect);
    
    if (embedImg.channels() == 4) {
        // Process transparent image
        vector<Mat> embedChannels;
        split(embedImg, embedChannels);
        
        Mat alpha;
        embedChannels[3].convertTo(alpha, CV_32FC1, 1.0/255.0);
        multiply(blendMask, alpha, blendMask);  // Combine masks
        
        vector<Mat> qrChannels;
        split(qrROI, qrChannels);
        
        // Create target mats
        Mat targetR(embedImg.size(), CV_32FC1, (topColor[0] + rightColor[0] + bottomColor[0] + leftColor[0])/4);
        Mat targetG(embedImg.size(), CV_32FC1, (topColor[1] + rightColor[1] + bottomColor[1] + leftColor[1])/4);
        Mat targetB(embedImg.size(), CV_32FC1, (topColor[2] + rightColor[2] + bottomColor[2] + leftColor[2])/4);
        
        // Convert channels
        Mat embedR, embedG, embedB;
        embedChannels[0].convertTo(embedR, CV_32FC1);
        embedChannels[1].convertTo(embedG, CV_32FC1);
        embedChannels[2].convertTo(embedB, CV_32FC1);
        
        // Blend operation
        Mat resultR = blendMask.mul(embedR) + (1 - blendMask).mul(targetR);
        Mat resultG = blendMask.mul(embedG) + (1 - blendMask).mul(targetG);
        Mat resultB = blendMask.mul(embedB) + (1 - blendMask).mul(targetB);
        
        // Convert back and merge
        resultR.convertTo(resultR, CV_8UC1);
        resultG.convertTo(resultG, CV_8UC1);
        resultB.convertTo(resultB, CV_8UC1);
        
        vector<Mat> mergedChannels = {resultR, resultG, resultB};
        Mat result;
        merge(mergedChannels, result);
        result.copyTo(qrROI);
    } else {
        // Process opaque image
        Mat embedFloat;
        embedImg.convertTo(embedFloat, CV_32FC3);
        
        // Create target color
        Vec3f targetColor(
            (topColor[0] + rightColor[0] + bottomColor[0] + leftColor[0])/4,
            (topColor[1] + rightColor[1] + bottomColor[1] + leftColor[1])/4,
            (topColor[2] + rightColor[2] + bottomColor[2] + leftColor[2])/4
        );
        
        Mat target(embedImg.size(), CV_32FC3, targetColor);
        
        // Blend operation
        vector<Mat> blendChannels;
        split(blendMask, blendChannels);
        blendChannels.resize(3);  // Duplicate for all channels
        
        Mat blended;
        multiply(embedFloat, blendMask, embedFloat);  // image * mask
        multiply(target, 1 - blendMask, target);      // target * (1-mask)
        add(embedFloat, target, blended);             // combined
        
        blended.convertTo(blended, CV_8UC3);
        blended.copyTo(qrROI);
    }
}

// void embedImage(const string& qrPath, const string& embedPath, const string& outputPath, const string& userRandomData = "") {
void embedImage(const string& qrPath, const string& embedPath, const string& outputPath, const string& userRandomData, int blendPercent){

    // Read images
    Mat qrImg = imread(qrPath, IMREAD_COLOR);
    if (qrImg.empty()) {
        cerr << "Error: Could not read QR code image" << endl;
        return;
    }

    Mat embedImg = imread(embedPath, IMREAD_UNCHANGED);
    if (embedImg.empty()) {
        cerr << "Error: Could not read image to embed" << endl;
        return;
    }

    // Calculate module size (simplified)
    int moduleSize = max(1, qrImg.cols / 21);

    // Resize embed image
    int maxSize = min(qrImg.rows, qrImg.cols) / 3;
    if (embedImg.rows > maxSize || embedImg.cols > maxSize) {
        double ratio = min(static_cast<double>(maxSize) / embedImg.rows, 
                         static_cast<double>(maxSize) / embedImg.cols);
        resize(embedImg, embedImg, Size(), ratio, ratio, INTER_AREA);
    }

    // Find random safe position
    // random_device rd;
    // mt19937 gen(rd());
    
    // Enhanced random seed generation
    random_device rd;
    string combinedSeed = to_string(rd()) + userRandomData;  // Combine both sources
    
    // Create a more complex hash of the combined seed
    size_t seedHash = 0;
    for (char c : combinedSeed) {
        seedHash = (seedHash << 5) + seedHash + c;  // Simple hash mixing
    }
    
    // Initialize random generators with different seeds derived from our hash
    mt19937 genX(seedHash);
    mt19937 genY(seedHash ^ 0x55555555);  // XOR with pattern for different sequence

    uniform_int_distribution<> distX(0, qrImg.cols - embedImg.cols);
    uniform_int_distribution<> distY(0, qrImg.rows - embedImg.rows);

    int x, y;
    int attempts = 0;
    const int maxAttempts = 100;

    do {
        x = distX(genX);
        y = distY(genY);
        attempts++;
    } while (!isSafeZone(qrImg, x, y, embedImg.cols, moduleSize) && attempts < maxAttempts);

    if (attempts >= maxAttempts) {
        cout << "Warning: Could not find optimal position" << endl;
        x = max(0, min(x, qrImg.cols - embedImg.cols));
        y = max(0, min(y, qrImg.rows - embedImg.rows));
    }

    // Blend images with edge matching
     blendEdgesWithQR(qrImg, embedImg, x, y, moduleSize,blendPercent);



    // Save result
    if (!imwrite(outputPath, qrImg)) {
        cerr << "Error: Could not save output image" << endl;
        return;
    }
    cout << "Success: Embedded image saved to " << outputPath << endl;
}

// int main() {
//     cout << "QR Code Image Embedder" << endl;
//     cout << "----------------------" << endl;

//     string qrPath, embedPath, outputPath, userRandomData;
    
//     cout << "Enter QR code image path: ";
//     getline(cin, qrPath);
    
//     cout << "Enter image to embed path: ";
//     getline(cin, embedPath);
    
//     cout << "Enter output path: ";
//     getline(cin, outputPath);

//     cout << "Enter random seed text (optional, press enter to skip): ";
//     getline(cin, userRandomData);

//     // Remove quotes if present
//     auto removeQuotes = [](string& s) {
//         s.erase(remove(s.begin(), s.end(), '\"'), s.end());
//     };
//     removeQuotes(qrPath);
//     removeQuotes(embedPath);
//     removeQuotes(outputPath);

//     embedImage(qrPath, embedPath, outputPath);

//     return 0;
// }

int safe_stoi(const std::string& s, int default_val) {  //edited: added this function
    try {
        return std::stoi(s);
    } catch (...) {
        return default_val;
    }
}

int main(int argc, char** argv) {
    if (argc < 6) {
        cerr << "Usage: " << argv[0] << " <qrPath> <embedPath> <outputPath> <randomSeed> <blendPercent>" << endl;
        return 1;
    }

    string qrPath = argv[1];
    string embedPath = argv[2];
    string outputPath = argv[3];
    string randomSeed = argv[4];
    // int blendPercent = stoi(argv[5]);  //edited: changed a bit
    int blendPercent = safe_stoi(argv[5], 30);  // default 30

    embedImage(qrPath, embedPath, outputPath, randomSeed, blendPercent);
    return 0;
}




